keps/754-time-dependent-job-priority/README.md

```diff
@@ -0,0 +1,197 @@
+nitty-gritty.
+-->
+
+### User Stories
+
+#### Story 1
+
+The job doesn't have a high priority but is intended for production, so it needs to start execution within three hours.
+
+
+### Risks and Mitigations
+
+If the job is just popped out for scheduling, the running job would be preempted easily by other Jobs.  
+That's why this mechanism should be based on priority of job.
+
+## Design Details
+
+### Kueue WorkloadPriorityClass API
+
+We introduce the Kueue `WorkloadPriorityClass` API.
+
+```golang
+// WorkloadPriorityClass defines basePriority and time dependent calculation logic
+type WorkloadPriorityClass struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   WorkloadPrioritySpec   `json:"spec,omitempty"`  
+}
+
+type WorkloadPrioritySpec struct {
+  // initial priority of workload
+	Priority int32 `json:"priority,omitempty"`
+
+  // +optional
+  DynamicPriorityPolicy DynamicPriorityPolicy `json:"dynamicPriorityPolicy,omitempty"`
+
+}
+
+type DynamicPriorityPolicy struct {
+
+  // step is a value that is added to the basePriority every time the duration defined by delayForStep elapses. 
+  Step int32 `json:"step,omitempty"`
+
+  // max defines the maximum number of Priority. We don't want it to go to infinity.
+  Max int32 `json:"max,omitempty"`
+
+  // delayForStep indicates the time to wait before changing the priority after a timeout
+  DeleyForStep time.Duration `json:"delayForStep,omitempty"`
+
+}
+
+```
+
+### How to use WorkloadPriorityClass on Job
+The `WorkloadPriorityClass` for a Job is defined using labels instead of `PriorityClass`.  
+We do not use the `spec.PriorityClass` field for `Job` because `Kueue` wouldn't manage `spec.PriorityClass`.
+
+```yaml
+# sample-job.yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  generateName: sample-job-
+  labels:
+    kueue.x-k8s.io/queue-name: user-queue
+    kueue.x-k8s.io/workload-priority-class: sample-priority
+```
+
+### Priority Calculation
+
+Priority is calculated based on the following equation:
+```
+Priority = min(priority + int(ceil(elapsedTime / delayForStep)) * step, max)
+```
+
+Let's calculate the priority for the following example:
+```yaml
+apiVersion: kueue.x-k8s.io/v1beta1
+kind: WorkloadPriorityClass
+metadata:
+  name: sample-priority
+Spec:
+  priority: 100
+  dynamicPriorityPolicy:
+    step: 400
+    max: 1000
+    delayForStep: 1h
+```
+
+Calculation:
+
+```
+Priority when the job is created: 100
+Priority after 1 hour: 100+400=500
+Priority after 2 hour: 100+400*2=900
+Priority after 3 hour: min(100+400*3, 1000)=1000
+...
+Priority after n hour: min(100+400*n, 1000)=1000
+```
+
+### Other points
+If a Job has both the `spec.PriorityClass` and WorkloadPriorityClass defined, `spec.PriorityClass` is ignored.  
+Since the `creationTimeStamp` of the Job is always available, the calculation can be performed every time
+without updating the values of Job and Workload Priority.
```

Comment:
This can make troubleshooting harder. In case of  a mix of jobs with different workload priority class it will be impossible to tell what the current order of workloads actually is at the moment (without copying them to a spreadsheet or running some additional custom scripts).